vtex-apps/render-runtime
# Feature/device blocks

Allows blocks from different devices to load lazily when the screen resizes. 
Incidentally, allows mobile blocks to not load on desktop and vice versa, improving the performance a little bit, and fixes an issue of missing blocks on certain devices/situations/tablet sizes.

### How to test
Open https://deviceblocks--alssports.myvtex.com/

1 - Via devtools, resize the viewport to mobile dimensions, and notice that the header will load lazily.
(there's an unrelated CSS bug that affects the content size in this example)
![desktop-mobile](https://user-images.githubusercontent.com/5691711/82061723-77143180-969f-11ea-9a94-0a1e3d84cbce.gif)

Doing the same on https://www.als.com/ displays the mobile header right away (which means it's already loaded on desktop, which might not be optimized)

2 - via devtools, on mobile mode, refresh the page. It should load the mobile header straight away. Resize it to desktop, and the desktop header should load lazily.

![mobile-desktop](https://user-images.githubusercontent.com/5691711/82062257-39fc6f00-96a0-11ea-80af-558385a9285a.gif)

Doing the same on https://www.als.com/ should either display the desktop header right away (which might not be optimized) or shouldn't display the desktop header at all (which is a bug)

3 - On both cases, you can see on the "Network" tab, on the `__pickRuntime` requests, that there's a `__device` field on the request, and that only the blocks for the respective devices come through. Also, on the initial `__RUNTIME__` value of the initial HTML request, the same should also apply.

### Depends on:
https://github.com/vtex/pages-graphql/pull/368
https://github.com/vtex/render-server/pull/615
https://github.com/vtex-apps/device-detector/pull/14
